invenio_rdm_migrator/streams/records/table_load.py

```diff
@@ -0,0 +1,209 @@
+                id=_pid_pk(),
+                pid_type="doi",
+                pid_value=rec["json"]["pids"]["doi"]["identifier"],
+                status="R",
+                object_type="rec",
+                object_uuid=rec["id"],
+                created=now,
+                updated=now,
+            )
+        # OAI
+        yield PersistentIdentifier(
+            id=_pid_pk(),
+            pid_type="oai",
+            pid_value=rec["json"]["pids"]["oai"]["identifier"],
+            status="R",
+            object_type="rec",
+            object_uuid=rec["id"],
+            created=now,
+            updated=now,
+        )
+
+        # parent
+        if parent["json"]["id"] not in self.parent_cache:
+            self.parent_cache[parent["json"]["id"]] = dict(
+                id=parent["id"],
+                version=dict(latest_index=rec["index"], latest_id=rec["id"]),
+            )
+            parent_pid = parent["json"]["pid"]
+            # record
+            yield RDMParentMetadata(
+                id=parent["id"],
+                json=parent["json"],
+                created=parent["created"],
+                updated=parent["updated"],
+                version_id=parent["version_id"],
+            )
+            # recid
+            yield PersistentIdentifier(
+                id=parent_pid["pk"],
+                pid_type=parent_pid["pid_type"],
+                pid_value=parent["json"]["id"],
+                status=parent_pid["status"],
+                object_type=parent_pid["obj_type"],
+                object_uuid=parent["id"],
+                created=now,
+                updated=now,
+            )
+        else:
+            # parent in cache - update version
+            cached_parent = self.parent_cache[parent["json"]["id"]]
+            # check if current record is a new version of the cached one
+            if cached_parent["version"]["latest_index"] < rec["index"]:
+                cached_parent["version"] = dict(
+                    latest_index=rec["index"], latest_id=rec["id"]
+                )
+
+    def prepare(self, output_dir, datagen, **kwargs):
+        """Compute rows."""
+        # use this context manager to close all opened files at once
+        with contextlib.ExitStack() as stack:
+            out_files = {}
+            for entry in datagen:
+                # is_db_empty would come in play and make _generate_pks optional                    
+                self._generate_pks(entry)
+                for entry in self._generate_db_tuples(entry):
+                    if entry._table_name not in out_files:
+                        fpath = output_dir / f"{entry._table_name}.csv"
+                        out_files[entry._table_name] = csv.writer(
+                            stack.enter_context(open(fpath, "w+"))
+                        )
+                    writer = out_files[entry._table_name]
+                    writer.writerow(as_csv_row(entry))
+
+        return self.tables
```

Comment:
minor: This part looks records-agnostic, so it could be moved up in the parent class.

(Note: it's my bad that the versions state equivalent is not relying on the same mechanism for figuring out which CSV file to write to, so it could be unified)

Reply:
100% agree!